firebase-common/src/main/java/com/google/firebase/components/ComponentRuntime.java

```diff
@@ -53,12 +57,50 @@ public ComponentRuntime(
     }
     Collections.addAll(componentsToAdd, additionalComponents);
 
-    components = Collections.unmodifiableList(ComponentSorter.sorted(componentsToAdd));
+    CycleDetector.detect(componentsToAdd);
 
-    for (Component<?> component : components) {
+    for (Component<?> component : componentsToAdd) {
       register(component);
     }
     validateDependencies();
+
+    processSetComponents();
+  }
+
+  /** Populates lazySetMap to make set components available for consumption via set dependencies. */
+  private void processSetComponents() {
+    Map<Class<?>, Set<Lazy<?>>> setIndex = new HashMap<>();
+    for (Map.Entry<Component<?>, Lazy<?>> entry : components.entrySet()) {
+      Component<?> component = entry.getKey();
+
+      // only process set components.
+      if (component.isValue()) {
+        continue;
+      }
+
+      Lazy<?> lazy = entry.getValue();
+
+      for (Class<?> anInterface : component.getProvidedInterfaces()) {
+        if (!setIndex.containsKey(anInterface)) {
+          setIndex.put(anInterface, new HashSet<>());
+        }
+        setIndex.get(anInterface).add(lazy);
+      }
+    }
+
+    for (Map.Entry<Class<?>, Set<Lazy<?>>> entry : setIndex.entrySet()) {
+      Set<Lazy<?>> lazies = entry.getValue();
+      lazySetMap.put(
+          entry.getKey(),
+          new Lazy<>(
+              () -> {
+                Set<Object> set = new HashSet<>();
+                for (Lazy<?> lazy : lazies) {
+                  set.add(lazy.get());
+                }
+                return Collections.unmodifiableSet(set);
```

Comment:
> Is there some kind of an addAll that we can use?

Hm, so to call addAll we need a Collection<T>, but we have a Collection<Lazy<T>>, now with streams this whole thing could be just a few lines of `map()`s and a collect but this is what we have to use I think.

> Also, consider a copyOnWrite set

I think it would be appropriate if we were optimizing for the use case where users are expected to modify this set as opposed to it being an edge case. Otherwise we are making everyone pay the synchronization cost of copyOnWrite sets which is not present in immutable collections. Also I think if some one really needs to modify the set, they can copy it explicitly, wdyt?

Reply:
Makes sense